lib/libzfs/libzfs_sendrecv.c

```diff
@@ -3592,7 +3592,7 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 		goto out;
 	}
 
-	if (top_zfs && *top_zfs == NULL)
+	if (top_zfs && (*top_zfs == NULL || strcmp(*top_zfs, name) == 0))
```

Comment:
Fix from https://github.com/zfsonlinux/zfs/commit/3cbe89b12a55a0ff350d7434834f4f48227896a3 requires the following bits from the first encryption commit (https://github.com/zfsonlinux/zfs/commit/b52563034230b35f0562b6f40ad1a00f02bd9a05):

```diff
diff --git a/lib/libzfs/libzfs_sendrecv.c b/lib/libzfs/libzfs_sendrecv.c
index c5acd21..cadf16c 100644
--- a/lib/libzfs/libzfs_sendrecv.c
+++ b/lib/libzfs/libzfs_sendrecv.c
@@ -3577,6 +3577,7 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
                }
 
                newfs = B_TRUE;
+               *cp = '/';
        }
 
        if (flags->verbose) {
```

This should resolve "receive-o-x_props_override" test failure.